### ISSUE TITLE
Documentation clarification

### ISSUE THREAD

**Issue by magic-lantern** [gt-health/GT-FHIR2#49]
I've installed the current version of gt-fhir2 as well as the sample OMOPv5 database on Postgresql with 5 sample patients.

Using both a fhir client (python) and the HAPI-FHIR test UI (http://<my_host>:8080/gt-fhir/tester/) I can only view resources. I cannot delete, add, nor update. So far, I've just tested with Patient and Encounter resources, but am not hopeful that others will be different. When I use the Tester, I get a HTTP 200 response, but the data isn't saved to the database, nor is it visible in the web UI.

I'm not sure if this is currently how the system is designed to work (read-only), if I'm doing something wrong, or if there is a known bug causing this problem.

Any guidance? I'm happy to provide tomcat logs, DB logs, more code, etc. if that's needed for troubleshooting.

**Comment by myungchoi** [comment]
Can you see list of Patients when you call the API? If you can only see list of resources but not the contents, then it could be because you haven't set the environment variable for DB endpoint and username/password.

**Comment by myungchoi** [comment]
> I've installed the current version of gt-fhir2 as well as the sample OMOPv5 database on Postgresql with 5 sample patients.
> 
> Using both a fhir client (python) and the HAPI-FHIR test UI (http://<my_host>:8080/gt-fhir/tester/) I can only view resources. I cannot delete, add, nor update. So far, I've just tested with Patient and Encounter resources, but am not hopeful that others will be different. When I use the Tester, I get a HTTP 200 response, but the data isn't saved to the database, nor is it visible in the web UI.
> 
> I'm not sure if this is currently how the system is designed to work (read-only), if I'm doing something wrong, or if there is a known bug causing this problem.
> 
> Any guidance? I'm happy to provide tomcat logs, DB logs, more code, etc. if that's needed for troubleshooting.

Can you see list of Patients when you call the API? If you can only see list of resources but not the contents, then it could be because you haven't set the environment variable for DB endpoint and username/password.

**Comment by magic-lantern** [comment]
Yes, I can see data for all resources I've tried - Patients, Encounters, Medications, Observations, etc.

In my initial attempt to setup my environment, I did have the DB enpoint information wrong, and in that case the HAPI Tester and API would not actually respond (but luckily the reason for the error did display in logs).

Here's an example of data returned from the Patient Resource:

```javascript
{
  "resourceType": "Patient",
  "id": "1",
  "text": {
    "status": "generated",
    "div": "<div xmlns=\"http://www.w3.org/1999/xhtml\"><div class=\"hapiHeaderText\">Byron <b>TEST1 </b></div><table class=\"hapiPropertyTable\"><tbody><tr><td>Identifier</td><td>111-222-3333</td></tr><tr><td>Address</td><td><span>886 Fox Valley Dr. </span><br/><span>null </span><br/><span>Stone Mountain </span><span>GA </span></td></tr><tr><td>Date of birth</td><td><span>09 December 1984</span></td></tr></tbody></table></div>"
  },
  "identifier": [
    {
      "value": "111-222-3333"
    }
  ],
  "active": true,
  "name": [
    {
      "family": "Test1",
      "given": [
        "Byron"
      ]
    }
  ],
  "gender": "male",
  "birthDate": "1984-12-09",
  "address": [
    {
      "use": "home",
      "line": [
        "886 Fox Valley Dr."
      ],
      "city": "Stone Mountain",
      "state": "GA",
      "postalCode": "30088"
    }
  ]
}
```

**Comment by myungchoi** [comment]
Do you have write permission to the DB that you set up with the username and password? Can you post tomcat log portion when you tried to delete the entry?

**Comment by magic-lantern** [comment]
I'm using the default omop_v5 account and database as created by the smart-postgresql script as documented in the README.md for this project `First clone the hdap-devops branch of the smart-platform-docker project`

I checked permissions, and it appears they are correct by default - the omop_v5 user has select, insert, and update, and delete privilges on 62 tables:

```
omop_v5=# select count(1) from information_schema.table_privileges where table_schema = 'omop_v5' and grantee = 'omop_v5' and privilege_type = 'DELETE';
 count
-------
    62
(1 row)

omop_v5=# select count(1) from information_schema.table_privileges where table_schema = 'omop_v5' and grantee = 'omop_v5' and privilege_type = 'UPDATE';
 count
-------
    62
(1 row)

omop_v5=# select count(1) from information_schema.table_privileges where table_schema = 'omop_v5' and grantee = 'omop_v5' and privilege_type = 'SELECT';
 count
-------
    62
(1 row)

omop_v5=# select count(1) from information_schema.table_privileges where table_schema = 'omop_v5' and grantee = 'omop_v5' and privilege_type = 'INSERT';
 count
-------
    62
(1 row)
```

Additionally, I manually logged in via `psql` and was able to insert and delete new rows into the person table.

In looking closer at my tomcat logs, I noticed a lot of errors on startup. There are matching errors in my postgresql log as well:

```
2018-11-13 00:32:44.783 UTC [28800] omop_v5@omop_v5 STATEMENT:
            alter table if exists provider
               add constraint FK42u18iywx49yo0mtq0rmokpnf
               foreign key (care_site_id)
               references care_site
2018-11-13 00:32:44.785 UTC [28800] omop_v5@omop_v5 LOG:  execute <unnamed>:
            alter table if exists provider
               add constraint FKl0ojpyjvqamr2fuvv65dneykp
               foreign key (gender_concept_id)
               references concept
2018-11-13 00:32:44.785 UTC [28800] omop_v5@omop_v5 ERROR:  there is no primary key for referenced table "concept"
2018-11-13 00:32:44.785 UTC [28800] omop_v5@omop_v5 STATEMENT:
            alter table if exists provider
               add constraint FKl0ojpyjvqamr2fuvv65dneykp
               foreign key (gender_concept_id)
               references concept
2018-11-13 00:32:44.786 UTC [28800] omop_v5@omop_v5 LOG:  execute <unnamed>:
            alter table if exists provider
               add constraint FKt1a1la0phwc8bd523eeoqdich
               foreign key (gender_source_concept_id)
               references concept
2018-11-13 00:32:44.786 UTC [28800] omop_v5@omop_v5 ERROR:  there is no primary key for referenced table "concept"
```

And a small matching snippet from the tomcat log:

```
2018-11-12 17:32:44.782 [localhost-startStop-1] DEBUG org.hibernate.SQL [SqlStatementLogger.java:94] 
    
    alter table if exists provider 
       add constraint FK42u18iywx49yo0mtq0rmokpnf 
       foreign key (care_site_id) 
       references care_site
2018-11-12 17:32:44.784 [localhost-startStop-1] WARN  o.h.t.s.i.ExceptionHandlerLoggedImpl [ExceptionHandlerLoggedImpl.java:27] GenerationTarget encountered exception accepting command : Error executing DDL "
    alter table if exists provider 
       add constraint FK42u18iywx49yo0mtq0rmokpnf 
       foreign key (care_site_id) 
       references care_site" via JDBC Statement
org.hibernate.tool.schema.spi.CommandAcceptanceException: Error executing DDL "
    alter table if exists provider 
       add constraint FK42u18iywx49yo0mtq0rmokpnf 
       foreign key (care_site_id) 
       references care_site" via JDBC Statement
	at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:67)
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.applySqlString(AbstractSchemaMigrator.java:559)
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.applySqlStrings(AbstractSchemaMigrator.java:504)
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.applyForeignKeys(AbstractSchemaMigrator.java:433)
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.performMigration(AbstractSchemaMigrator.java:249)
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.doMigration(AbstractSchemaMigrator.java:114)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:183)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:72)
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:310)
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:467)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:939)
	at org.hibernate.jpa.HibernatePersistenceProvider.createContainerEntityManagerFactory(HibernatePersistenceProvider.java:149)
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:365)
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:388)
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:377)
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.afterPropertiesSet(LocalContainerEntityManagerFactoryBean.java:341)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1769)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1706)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:583)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:502)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:228)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:310)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1085)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:858)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:409)
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:291)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:103)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4776)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5240)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:754)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:730)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:734)
	at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:985)
	at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1856)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: org.postgresql.util.PSQLException: ERROR: there is no primary key for referenced table "care_site"
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2422)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2167)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:306)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:307)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:293)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:270)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:266)
	at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:291)
	at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:291)
	at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:54)
	... 42 common frames omitted
```

So far, the above startup errors seem the most likely candidate for the inability to delete and update records.

I'm not seeing any errors when I actually attempt to delete a patient though.

Attached is the full error log - its about 3.5MB, so it's pretty large.

[catalina.txt](https://github.com/gt-health/GT-FHIR2/files/2574162/catalina.txt)

**Comment by shwethashiva** [comment]
Hi magic-lantern,
I am trying to install the GT-FHIR2 in my machine. I took the sample_data from github repo and added the dump into my local postgresDb , followed the same procedure as in smart-postgresql. I then started application GT-FHIR2 using docker compose up. After all this If I hit the UI "http://localhost:9090/gt-fhir/tester I am getting 404 error. Could you please guide me on steps you followed. It will be very helpful. Thank you.

Regards,
Shwetha

**Comment by shwethashiva** [comment]
@magic-lantern Yes, this is no way related to this issue you are facing . I will add it as new issue.(PS: this is not a bug but indeed my incompetence regarding docker)
I am too adding localhost for the DB environment variables , how else to specify the postgres that is running in my localhost? Sorry if this sounds too basic. Appreciate your help.

**Comment by myungchoi** [comment]
@magic-lantern Ooops, yeah primary key needs to be specified. I didn't realize that our sample database does not have that set up. I am using my own database for testing, which has all the primary set up. I will regenerate the sample database. In the meantime, you can set primary key to all "id" column of the tables.

And, cascade delete could also cause the problem. Currently, we are not doing the cascade delete to prevent accidental deletes. If you want to delete, you need to delete a resource that the resource is referring to. However, if the resource you are about to delete is also referenced by another resource, you won't be able to delete as hibernate flag an error.

**Comment by myungchoi** [comment]
@magic-lantern For update, can you post the resource that you are trying to update? I will run it on my side to see if I can regenerate the issue.

**Comment by myungchoi** [comment]
> Hi magic-lantern,
> I am trying to install the GT-FHIR2 in my machine. I took the sample_data from github repo and added the dump into my local postgresDb , followed the same procedure as in smart-postgresql. I then started application GT-FHIR2 using docker compose up. After all this If I hit the UI "http://localhost:9090/gt-fhir/tester I am getting 404 error. Could you please guide me on steps you followed. It will be very helpful. Thank you.
> 
> Regards,
> Shwetha

@shwethashiva If you are using Mac OSX, you can't use localhost from inside the Docker container. Please see https://forums.docker.com/t/understanding-the-docker-for-mac-localhost-behavior/41921/3   Your environment setup should use an hostname that can be accessed.

**Comment by magic-lantern** [comment]
@myungchoi Thanks for your help and diligence on this issue. 

I ran the following SQL to add a bunch of indexes to my local postgresql instance:

```sql
select 'ALTER TABLE ' || tablename || ' ADD CONSTRAINT ' || tablename || '_id PRIMARY KEY (' || tablename || '_id);' FROM pg_catalog.pg_tables where tableowner = 'omop_v5' and tablename not like '%aud' and tablename not like 'smart%' order by tablename;
-- copy and paste output of the above statement into psql
-- there are a few warnings thrown as not all of these indexes can be created, but it gets the job done.

-- this existing index prevents creating the PRIMARY KEY index
drop index if exists f_person_person_id_idx;
ALTER TABLE f_person ADD CONSTRAINT f_person_id PRIMARY KEY (person_id);
```

Once I did the above, I still am unable to create new Patients, nor delete items that do not appear to have any dependencies (didn't try any updates).

Looking back at my Tomcat logs, I see 10 errors about a the view f_observation_view when Tomcat starts. Here's an example:

```
org.hibernate.tool.schema.spi.CommandAcceptanceException: Error executing DDL "
    alter table if exists f_observation_view
       add constraint FK83xuk8ouegfdgdm8vrd89k7du
       foreign key (source_concept_id)
       references concept" via JDBC Statement
        at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:67)
 ...
Caused by: org.postgresql.util.PSQLException: ERROR: "f_observation_view" is not a table or foreign table
```

In Postgresql views cannot have foreign key constraints it seems. 

In your local copy of the database, what is your definition of f_observation_view?

**Comment by magic-lantern** [comment]
Sorry @myungchoi , I realized I forgot to answer your other question about an example of an operation I'm attempting.

Using the "CRUD Operations" tab, I've tried to create, update and delete various items, and all I ever get is an HTTP 200 response - but not changes to my postgresql database.

Here's an example of a new patient I'm attempting to create:

```javascript
{
    "resourceType" : "Patient",
    "identifier": [
        {
            "value": "4242424242"
        }
    ],
    "active": true,
    "name": [
        {
            "family": "Test42",
            "given": [
                "Byron42"
            ]
        }
    ],
    "gender": "male",
    "birthDate": "1984-12-09",
    "address": [
        {
            "use": "home",
            "line": [
                "886 Fox Valley Dr."
            ],
            "city": "Stone Mountain",
            "state": "GA",
            "postalCode": "30088"
        }
    ]
}
```

I've also tried using the [SMART Python client](http://docs.smarthealthit.org/client-py/) and just plain Postman to manually make REST calls, but get the same result- no error returned, but no data actually changed in my database.

**Comment by myungchoi** [comment]
@magic-lantern Sorry for the delay. Until now, I couldn't find out your problem. But, I just realized that the web.xml file posted on the repository has readOnly option set top True. Please open the web.xml file in WEB-INF directory. And, find <init-param> with perm-name equal to readOnly. Set the value to False. The repository's default value is set to True to protect the database from being altered.

This should allow you to post the resource for writing.

**Comment by myungchoi** [comment]
@magic-lantern I made the server to throw exception when post/put is requested during the server at readOnly mode instead of absorbing silently. In this way, client will know that server is not accepting the method. Hope this resolves your issue.

**Comment by magic-lantern** [comment]
Thanks for the explanation and code change. I confirmed that an error message is now thrown when an update is attempted (tried to create a patient). That is much better!

I also was able to modify my web.xml to set readOnly to false. After that, I was able to add a patient.

Last question: How would a new user of GT-FHIR2 know that read-only is the default? I didn't see any documentation about that state in the README.md nor on the website. If I missed it, guess I need new glasses. If there is not any information about making it writeable, I recommend updating the main README.md.
